android/app/src/main/java/com/epfl/dedis/hbt/ui/register/ScanPassportFragment.kt

```diff
@@ -0,0 +1,142 @@
+import com.epfl.dedis.hbt.ui.MainActivity
+import com.epfl.dedis.hbt.ui.wallet.ImageAnalyzerProvider
+import com.epfl.dedis.hbt.utility.json.JsonService
+import com.google.mlkit.vision.text.TextRecognition
+import com.google.mlkit.vision.text.TextRecognizer
+import com.google.mlkit.vision.text.latin.TextRecognizerOptions
+import dagger.hilt.android.AndroidEntryPoint
+import java.util.regex.Pattern
+import javax.inject.Inject
+
+@AndroidEntryPoint
+class ScanPassportFragment : Fragment() {
+
+    companion object {
+        private val LINE_1_PATTERN = Pattern.compile("(P[A-Z0-9<]{1})([A-Z]{3})([A-Z0-9<]{39})")
+        private val LINE_2_PATTERN =
+            Pattern.compile("([A-Z0-9<]{9})([0-9]{1})([A-Z]{3})([0-9]{6})([0-9]{1})([M|F|X|<]{1})([0-9]{6})([0-9]{1})([A-Z0-9<]{14})([0-9<]{1})([0-9]{1})")
+    }
+
+    private var _binding: FragmentPassportScanBinding? = null
+
+    @Inject
+    lateinit var jsonService: JsonService
+
+    @Inject
+    lateinit var imageAnalyzerProvider: ImageAnalyzerProvider
+
+    // This property is only valid between onCreateView and
+    // onDestroyView.
+    private val binding get() = _binding!!
+
+    private var scanner: TextRecognizer? = null
+
+    override fun onCreateView(
+        inflater: LayoutInflater,
+        container: ViewGroup?,
+        savedInstanceState: Bundle?
+    ): View {
+        // Create a request permission launcher which will ask for permission when launched
+        val requestPermissionLauncher = registerForActivityResult(
+            ActivityResultContracts.RequestPermission(),
+            requireActivity().activityResultRegistry
+        ) {
+            // This is the callback of the permission granter
+            applyPermissionToView()
+        }
+
+        _binding = FragmentPassportScanBinding.inflate(inflater, container, false).apply {
+            requestPermissionButton.setOnClickListener {
+                // Launch the permission request on click
+                requestPermissionLauncher.launch(
+                    Manifest.permission.CAMERA
+                )
+            }
+        }
+
+        return binding.root
+    }
+
+    override fun onResume() {
+        super.onResume()
+        applyPermissionToView()
+    }
+
+    override fun onDestroy() {
+        super.onDestroy()
+
+        scanner?.close()
+    }
+
+    private fun startCamera() {
+        val cameraController = LifecycleCameraController(requireContext())
+        val previewView: PreviewView = binding.previewView
+
+        scanner = TextRecognition.getClient(TextRecognizerOptions.DEFAULT_OPTIONS)
+
+        cameraController.setImageAnalysisAnalyzer(
+            ContextCompat.getMainExecutor(requireActivity()),
+            imageAnalyzerProvider.provide(
+                scanner!!,
+                CameraController.COORDINATE_SYSTEM_VIEW_REFERENCED,
+                ContextCompat.getMainExecutor(requireActivity())
+            ) {
+                val raw = it?.text ?: return@provide
```

Comment:
I must say, I'm less impressed by Kotlin when I see this hieroglyphics programming ;-) It's like if a Perl programmer has infiltrated the Kotlin dev team...

Reply:
Its a convinient way to handle nullables.

Otherwise, it would look this way :
```
if (it == null) return@provide
if (it.text == null) return@provide

val raw = it.text
```

Which is a little longer

Reply:
Thanks for this old-man readable version :-) it does help.